README.md
<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/TAUR)](https://cran.r-project.org/package=TAUR)
[![R-CMD-check](https://github.com/quanteda/TAUR/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/quanteda/TAUR/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

# Companion package for _Text Analysis Using R_

Package intended to provide companion data and functions for _Text Analysis Using R_.

## How to Install

You can download the files and build the package from source, or you can install the package directly from GitHub as follows:

```r
remotes::install_github("quanteda/TAUR")
```

## Available corpora

Corpora contained in the package are the following:

Corpus | Name
--|--
20k TripAdvisor hotel reviews | data_corpus_TAhotels
2020 US Presidential Debates | data_corpus_debates
Donald Trump's tweets, 2009-2021 | data_corpus_trumptweets
Amicus curiae briefs from Bakke (1978) and Bollinger (2008) | data_corpus_amicus
Annual budget speeches from the Irish Dáil, 2008-2012 | data_corpus_irishbudgets
UK news articles from 2014 that mention immigration | data_corpus_immigrationnews
US State of the Union addresses from 1790 to present | data_corpus_sotu
UK political party manifestos, 1945-2005 | data_corpus_ukmanifestos

## Available dictionaries

Corpora contained in the package are the following:

Dictionary | Name
--|--
Laver and Garry (2001) | data_dictionary_LaverGarry
Moral Foundations Dictionary 2.0 | data_dictionary_MFD
Martindale's Regressive Image Dictionary | data_dictionary_RID



NEWS.md
# TAUR 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library("testthat")
library("quanteda")

test_check("TAUR")


tests/testthat/test-data.R
test_that("corpus data objects are current", {
    expect_true(is.corpus(data_corpus_amicus))
    expect_true(is.corpus(data_corpus_LMRD))
    expect_true(is.corpus(data_corpus_sotu))
    expect_true(is.corpus(data_corpus_ukmanifestos))
    expect_true(is.corpus(data_corpus_debates))
    expect_true(is.corpus(data_corpus_TAhotels))
    expect_true(is.corpus(data_corpus_trumptweets))
    expect_true(is.corpus(data_corpus_debates))
})

test_that("dictionary data objects are current", {
    expect_true(is.dictionary(data_dictionary_LaverGarry))
    expect_true(is.dictionary(data_dictionary_MFD))
    expect_true(is.dictionary(data_dictionary_RID))
})


tests/data_creation/create_debates.R
library("quanteda")
library("quanteda")
library("rvest")
library("stringr")
library("quanteda.tidy")

# assign URL of debates to an object called url_debates
url_debates <- "https://www.presidency.ucsb.edu/advanced-search?field-keywords=&field-keywords2=&field-keywords3=&from%5Bdate%5D=01-01-2020&to%5Bdate%5D=12-31-2020&person2=200301&category2%5B%5D=64&items_per_page=50"

source_page <- read_html(url_debates)

# get debate meta-data

nodes_pres <- ".views-field-title a"
text_pres <- ".views-field-field-docs-start-date-time-value.text-nowrap"

debates_meta <- data.frame(
    location = html_text(html_nodes(source_page, nodes_pres)),
    date =  html_text(html_nodes(source_page, text_pres)),
    stringsAsFactors = FALSE
)

# format the date
debates_meta$date <- as.Date(trimws(debates_meta$date),
                             format = "%b %d, %Y")

# get debate URLs
debates_links <- source_page |>
    html_nodes(".views-field-title a") |>
    html_attr(name = "href")

# add first part of URL to debate links
debates_links <- paste0("https://www.presidency.ucsb.edu", debates_links)

# scrape search results
debates_scraped <- lapply(debates_links, read_html)

# get character vector, one element per debate
debates_text <- sapply(debates_scraped, function(x) {
    html_nodes(x, "p") |>
        html_text() |>
        paste(collapse = "\n\n")
})

debates_meta$location <- factor(str_remove_all(debates_meta$location,
                                        "Presidential Debate at "))


# remove end of documents

remove_end_20200929 <- "\n\nDonald J. Trump, Presidential Debate at Case Western Reserve University in Cleveland, Ohio Online by Gerhard Peters and John T. Woolley, The American Presidency Project https://www.presidency.ucsb.edu/node/343824\n\nThe American Presidency ProjectJohn Woolley and Gerhard PetersContact\n\nTwitter Facebook\n\nCopyright © The American Presidency ProjectTerms of Service | Privacy | Accessibility\n\n"
remove_end_20201022 <- "\n\nDonald J. Trump, Presidential Debate at Belmont University in Nashville, Tennessee Online by Gerhard Peters and John T. Woolley, The American Presidency Project https://www.presidency.ucsb.edu/node/345898\n\nThe American Presidency ProjectJohn Woolley and Gerhard PetersContact\n\nTwitter Facebook\n\nCopyright © The American Presidency ProjectTerms of Service | Privacy | Accessibility\n\n"

dat_debates_text <- data.frame(text = debates_text)

dat_debates_text <- dat_debates_text |>
    mutate(text = gsub(".*News\\)\\n\\n", "", text), # remove start of string that does not relate to debate
           text = str_remove_all(text, remove_end_20200929),
           text = str_remove_all(text, remove_end_20201022),
           text = str_remove_all(text, "\\|\\|"),
           text = str_replace_all(text, " {0,1}(\\. {0,1}){3,4}", "..."), # remove spaces in ellipses
           text = str_replace_all(text, fixed(":..."), ": ..."),          # for spaces removed after colons
           text = str_replace_all(text, "\\-\\n", "...\n"),               # use ... for end of line not -
           text = str_replace_all(text, fixed("- ["), "... ["),           # use ... for internal interruption
           text = str_replace_all(text, " {0,1}— {0,1}\\n", "...\n"),
           text = str_replace_all(text, fixed("spewting"), "spewing"),
           text = str_replace_all(text, fixed("200--250"), "200-250"),
           text = str_replace_all(text, fixed("Every body"), "Everybody"),
           text = str_replace_all(text, "(\\w) {0,1}-- {0,1}(\\w)", "\\1 -- \\2"))

dat_debates_text <- dat_debates_text |>
  mutate(text = str_replace_all(text, " -...\\n", "...\n"))

data_corpus_debates <- corpus(dat_debates_text$text,
                              docvars = debates_meta)



# check that irrelevant text at start and end has been removed
data_corpus_debates
as.character(data_corpus_debates[1])
as.character(data_corpus_debates[2])


docnames(data_corpus_debates) <- paste0("Debate ", data_corpus_debates$date)


# inspect
quanteda.tidy::glimpse(data_corpus_debates)

meta(data_corpus_debates) <- list(
  description = "2020 US Presidential Debates between Donald J. Trump and Joe Biden",
  source = "The Presidential Presidency Project",
  url = "https://www.presidency.ucsb.edu",
  author = "UC Santa Barbara",
  keywords = c("political", "US politics", "United States", "presidents", "Trump", "Biden", "televised debates"),
  title = "2020 US Presidential Debates"
)

usethis::use_data(data_corpus_debates, overwrite = TRUE)


R/data-dictionaries.R
#' Martindale's Regressive Imagery Dictionary (RID)
#'
#' The English Regressive Imagery Dictionary (RID) consists of about 3,150 words
#' and roots assigned to 29 categories of primary process cognition, 7
#' categories of secondary process cognition, and 7 categories of emotions,
#' designed by Martindale (1975, 1990) to measure primordial vs. conceptual
#' thinking.
#' @format The dictionary has three levels of nesting, and forms 43 keys in
#'   total and a total of 3,150 values.
#' @references Martindale, C. (1975). *Romantic Progression: The Psychology
#' of Literary History.* Washington, D.C.: Hemisphere.
#'
#' Martindale, C. (1990). *The Clockwork Muse: The Predictability of
#' Artistic Change.* New York: Basic Books.
#' @keywords data
"data_dictionary_RID"

#' Moral Foundations Dictionary
#'
#' @description A \pkg{quanteda} [dictionary][quanteda::dictionary] object containing
#' the Moral Foundations Dictionary, a publicly available dictionaries with
#' information on the proportions of virtue and vice words for each foundation.
#' The categories are harm (vice/virtue), fairness (vice/virtue), ingroup (vice/virtue),
#' authority (vice/virtue), purity (vice/virtue) and morality (general).
#'
#' @description This version is 2.0 version of the dictionary,
#'   [recommended](http://www.jeremyfrimer.com/uploads/2/1/2/7/21278832/summary.pdf)
#'   over the first version of the MDF by its authors.
#' @source http://www.jeremyfrimer.com/research-downloads.html; a previous
#'   version is available at <http://moralfoundations.org/othermaterials>
#' @references
#'   Frimer, J. et. al. (2017).  Moral Foundations Dictionaries for
#'   Linguistic Analyses, 2.0. University of Winnipeg Manuscript.
#'   <http://www.jeremyfrimer.com/uploads/2/1/2/7/21278832/summary.pdf>.
#'
#'   Haidt, J., Graham, J., & Nosek, B.A. (2009). [Liberals and Conservatives
#'   Rely on Different Sets of Moral
#'   Foundations](http://dx.doi.org/10.1037/a0015141). *Journal of Personality
#'   and Social Inquiry*, 20(2--3), 110--119.
#'
#'   Graham, J., & Haidt, J. (2016). [Moral Foundations
#'   Dictionary.](http://moralfoundations.org/othermaterials):
#'   <http://moralfoundations.org/othermaterials>.
#' @keywords data
"data_dictionary_MFD"

#' Laver and Garry Dictionary of Policy Positions
#'
#' A \pkg{quanteda} [dictionary][quanteda::dictionary] object containing the Laver and Garry
#' dictionary which been developed to estimates the policy positions of political actors
#' in the United Kingdom by comparing their speeches and written documents to key words
#' found in the British Conservative and Labour manifestos of 1992. Note: Please remember
#' that this dictionary was customized to reflect the policy positions of UK political parties.
#' @format The dictionary has two levels of nesting with 7 main policy areas (level 1) divided up into 19
#' sub-categories (level 2).
#' @source
#'   <https://provalisresearch.com/products/content-analysis-software/wordstat-dictionary/laver-garry-dictionary-of-policy-position/>
#'
#' @references
#'   Laver. M. & Garry, J. (2000). [Estimating Policy Positions
#'   from Political Texts](https://doi.org/10.2307/2669268). *American Journal of Political Science*, 44(3), 619--634.
#' @keywords data
"data_dictionary_LaverGarry"


R/data-corpora.R
#' Amicus curiae briefs from Bakke (1978) and Bollinger (2008)
#'
#' Texts of amicus curiae briefs labelled as being either pro-petitioner or
#' pro-respondent in US Supreme court cases on affirmative action, Bakke (1978)
#' and Bollinger (2008), taken from Evans et al (2007).
#' @format
#'   The corpus consists of 102 documents and includes the following document
#'   variables: \describe{
#'   \item{trainclass}{A character indicating the class of 4 documents that can
#'   be used to train a classifier (the remaining ones are NA); P:
#'   Pro-Petitioner; R: Pro-Respondent.}
#'   \item{testclass}{A character indicating the "true" class of 98 remaining
#'   documents to assess the prediction of the classification; AP:
#'   Pro-Petitioner; AR: Pro-Respondent.} }
#' @references Evans, M., McIntosh, W., Lin, J., & Cates, C. (2007).
#'   "[Recounting the Courts? Applying Automated Content Analysis to Enhance
#'   Empirical Legal
#'   Research](https://doi.org/10.1111/j.1740-1461.2007.00113.x)." *Journal of
#'   Empirical Legal Studies* 4(4): 1007--1039.
#' @keywords data
"data_corpus_amicus"

#' U.S. State of the Union addresses from 1790 to present
#'
#' A corpus object of every US State-of-the-Union address, from 1790 to present.
#' Where interjections or records of audience reaction occurred, such as
#' "(Applause.)", these have been removed.
#' @format
#'   The corpus includes the following document variables: \describe{
#'   \item{FirstName}{President's first and middle names.}
#'   \item{President}{President's last name.}
#'   \item{Date}{Date of the delivery of the speech or document.}
#'   \item{delivery}{Either `written` or `spoken`, depending on the
#'   format. See Source.}
#'   \item{type}{Either `SOTU` for an official State of the Union Address,
#'   or `other`, for a different form of speech or message. See Source.}
#'   \item{party}{President's party.}
#'   }
#' @source The American Presidency Project,
#'   <http://www.presidency.ucsb.edu/sou.php>.
"data_corpus_sotu"

#' UK political party manifestos, 1945--2005
#'
#' A corpus object containing 105 UK Manifestos from 1945--2005, with party and year attributes.
#' @format
#'   The corpus includes the following document variables: \describe{
#'   \item{Country}{Country name (UK).}
#'   \item{Type}{Character indicating the type of election: national (`natl`) or regional (`regl`.}
#'   \item{Language}{Language (`en`).}
#'   \item{Party}{A character abbreviation of the political party.}
#'   }
"data_corpus_ukmanifestos"

#' Large Movie Review Dataset from Maas et. al. (2011)
#'
#' A corpus object containing a dataset for sentiment classification containing
#' 25,000 highly polar movie reviews for training, and 25,000 for testing, from
#' Maas et. al. (2011).
#' @format The corpus docvars consist of:
#'   \describe{
#'   \item{docnumber}{serial (within set and polarity) document number}
#'   \item{rating}{user-assigned movie rating on a 1-10 point integer scale}
#'   \item{set}{used for test v. training set}
#'   \item{polarity}{either `neg` or `pos` to indicate whether the
#'     movie review was negative or positive.  See Maas et al (2011) for the
#'     cut-off values that governed this assignment.}
#'   }
#' @references
#' Andrew L. Maas, Raymond E. Daly, Peter T. Pham, Dan Huang, Andrew Y. Ng, and
#' Christopher Potts. (2011). "[Learning Word Vectors for Sentiment
#' Analysis](http://ai.stanford.edu/~amaas/papers/wvSent_acl2011.pdf)". The 49th
#' Annual Meeting of the Association for Computational Linguistics (ACL 2011).
#' @source <http://ai.stanford.edu/~amaas/data/sentiment/>
#' @keywords data
"data_corpus_LMRD"

#' Trump Twitter archive
#'
#' A corpus object containing a dataset consisting of all of Trump's Tweets,
#' going back to 2009.  Total: 56,571.
#' @format The docname is the ID of the original Tweet.  The corpus docvars consist of:
#'   \describe{
#'   \item{isRetweet}{logical; if *TRUE*, the Tweet was a retweet.}
#'   \item{isDeleted}{logical; if *TRUE*, the Tweet was deleted.}
#'   \item{device}{factor; the device from which the Tweet was sent.}
#'   \item{favorites}{integer; the total number of favorites the Tweet received.}
#'   \item{retweets}{integer; total number of times the Tweet was retweeted.}
#'   \item{isFlagged}{logical; if *TRUE*, the Tweet was flagged.}
#'   \item{datetime}{datetime format in UTC.}
#'   }
#' @source <https://www.thetrumparchive.com>
#' @keywords data
"data_corpus_trumptweets"

#' TripAdvisor Hotel Reviews
#'
#' A corpus object containing 20k TripAdvisor hotel reviews, and ratings.
#' @format The only corpus docvar consists of `Rating`, ranging from 1 to 5.
#' @source <https://www.kaggle.com/andrewmvd/trip-advisor-hotel-reviews>
#' @references Alam, M. H., Ryu, W.-J., Lee, S. (2016). Joint multi-grain topic
#'   sentiment: modeling semantic aspects for online reviews. _Information
#'   Sciences_ 339, 206–223. \doi{10.1016/j.ins.2016.01.013}
#' @section License: [Attribution-NonCommercial 4.0 International (CC BY-NC
#'   4.0)](https://creativecommons.org/licenses/by-nc/4.0/)
#' @keywords data
"data_corpus_TAhotels"


#' 2020 US Presidential Debates
#'
#' A corpus object containing the transcripts of the two US Presidential Debates
#' Donald J. Trump and Joe Biden. Each document includes the full debate text by the two
#' politicians and the moderator.
#' @format The docname is the date of the debate.  The corpus docvars consist of:
#'   \describe{
#'   \item{location}{factor; location of the debate (Cleveland and Nashville).}
#'   \item{date}{date; date of the debate (2020-09-29 and 2020-10-22).}
#'   }
#' @source <https://www.presidency.ucsb.edu>
#' @references "The American Presidency Project"
#' @keywords data
#' @examples
#' # transform debate-level corpus to level of statements
#' library(quanteda)
#' data_corpus_debatesseg <- corpus_segment(data_corpus_debates,
#'                                          pattern =  "\\s*[[:upper:]]+:\\s+",
#'                                          valuetype = "regex",
#'                                          case_insensitive = FALSE)
"data_corpus_debates"
